File: Leetcode/Easy/remove_duplicates_from_sorted_list.cpp
/*
https://leetcode.com/problems/remove-duplicates-from-sorted-list/
*/

# include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    struct ListNode {
        int val;
        ListNode *next;
        ListNode() : val(0), next(nullptr) {}
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
 };

    ListNode* deleteDuplicates(ListNode* head) {
        
        ListNode* curr_node = head;
        ListNode* next_node;        
        
        while (curr_node != NULL && curr_node -> next != NULL) {
            
            next_node = curr_node -> next;
            
            if (next_node -> val == curr_node -> val) 
                curr_node -> next = next_node -> next;
            
            else
                curr_node = curr_node -> next;               
        }
        
        return head;        
    }
};

File: Data-Structures-Algorithms/Linked-Lists/Singly-Linked-Lists/remove-duplicates-from-sorted-list.java
// https://leetcode.com/problems/remove-duplicates-from-sorted-list/description/

// Given the head of a sorted linked list, delete all duplicates such that each element appears only once. Return the linked list sorted as well.

 

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {

        ListNode curr = head;
        ListNode next;

        while(curr != null && curr.next != null) {
            next = curr.next;

            if (curr.val == next.val) {
                curr.next = next.next;
            } else {
                curr = curr.next;
            }
        } 

        return head;
        
    }
}

